test/ExecutionFramework.cpp

```diff
@@ -50,25 +55,294 @@ string getIPCSocketPath()
+		m_vm(_vm)
+	{
+		if (_evmVersion == langutil::EVMVersion::homestead())
+			m_evmVersion = EVMC_HOMESTEAD;
+		else if (_evmVersion == langutil::EVMVersion::tangerineWhistle())
+			m_evmVersion = EVMC_TANGERINE_WHISTLE;
+		else if (_evmVersion == langutil::EVMVersion::spuriousDragon())
+			m_evmVersion = EVMC_SPURIOUS_DRAGON;
+		else if (_evmVersion == langutil::EVMVersion::byzantium())
+			m_evmVersion = EVMC_BYZANTIUM;
+		else if (_evmVersion == langutil::EVMVersion::constantinople())
+			m_evmVersion = EVMC_CONSTANTINOPLE;
+		else if (_evmVersion == langutil::EVMVersion::petersburg())
+			m_evmVersion = EVMC_PETERSBURG;
+		else
+			BOOST_REQUIRE(false);
+	}
+
+	struct Account
+	{
+		evmc_uint256be balance = {};
+		size_t nonce = 0;
+		bytes code;
+		evmc_bytes32 codeHash = {};
+		std::map<evmc_bytes32, evmc_bytes32> storage;
+	};
+
+	struct State
+	{
+		std::map<evmc_address, Account> accounts;
+		std::vector<LogEntry> logs;
+	};
+
+	Account* account(evmc_address const& _address)
+	{
+		auto it = m_state.accounts.find(_address);
+		return it == m_state.accounts.end() ? nullptr : &it->second;
+	}
+
+	void reset() { m_state = State{}; m_currentAddress = {}; }
+
+	bool account_exists(evmc_address const& _addr) noexcept final
+	{
+		return account(_addr) != nullptr;
+	}
+
+	evmc_bytes32 get_storage(evmc_address const& _addr, evmc_bytes32 const& _key) noexcept final
+	{
+		if (Account* acc = account(_addr))
+			return acc->storage[_key];
+		return {};
+	}
+
+	evmc_storage_status set_storage(
+		evmc_address const& _addr,
+		evmc_bytes32 const& _key,
+		evmc_bytes32 const& _value
+	) noexcept final
+	{
+		evmc_bytes32 previousValue = m_state.accounts[_addr].storage[_key];
+		m_state.accounts[_addr].storage[_key] = _value;
+
+		// TODO there are more possible values, especially "MODIFIED_AGAIN".
+		return previousValue == _value ? EVMC_STORAGE_UNCHANGED : EVMC_STORAGE_MODIFIED;
+	}
+
+	evmc_uint256be get_balance(evmc_address const& _addr) noexcept final
+	{
+		if (Account const* acc = account(_addr))
+			return acc->balance;
+		return {};
+	}
+
+	size_t get_code_size(evmc_address const& _addr) noexcept final
+	{
+		if (Account const* acc = account(_addr))
+			return acc->code.size();
+		return 0;
+	}
+
+	evmc_bytes32 get_code_hash(evmc_address const& _addr) noexcept final
+	{
+		if (Account const* acc = account(_addr))
+			return acc->codeHash;
+		return {};
+	}
+
+	size_t copy_code(
+		evmc_address const& _addr,
+		size_t _codeOffset,
+		uint8_t* _bufferData,
+		size_t _bufferSize
+	) noexcept final
+	{
+		// TODO is this supposed to fill with zero bytes?
+		Account const* acc = account(_addr);
+		if (!acc)
+			return 0;
+		size_t i = 0;
+		for (; i < _bufferSize && _codeOffset + i < acc->code.size(); i++)
+			_bufferData[i] = acc->code[_codeOffset + i];
+		return i;
+	}
+
+	void selfdestruct(evmc_address const& _addr, evmc_address const& _beneficiary) noexcept final
+	{
+		m_state.accounts[_beneficiary].balance = m_state.accounts[_addr].balance;
```

Comment:
There is a special case where `_beneficiary` is `_addr`.

Reply:
So the balance is not destroyed in that case?

Reply:
I think this implementation is ok unless you will call the destroyed account.